zerver/migrations/0064_add_file_name_field_to_realm_emoji.py

```diff
@@ -0,0 +1,171 @@
+            os.makedirs(dirname)
+
+    def write_local_file(self, path, file_data):
+        # type: (Text, binary_type) -> None
+        self.mkdirs(path)
+        with open(path, 'wb') as f:   # type: ignore
+            f.write(file_data)
+
+    def upload_files(self, response, resized_image, dst_path_id):
+        # type: (Response, binary_type, Text) -> None
+        dst_file = os.path.join(settings.LOCAL_UPLOADS_DIR, 'avatars', dst_path_id)
+        if resized_image:
+            self.write_local_file(dst_file, resized_image)
+        else:
+            self.write_local_file(dst_file, response.content)
+        self.write_local_file('.'.join((dst_file, 'original')), response.content)
+
+
+class S3Uploader(Uploader):
+    def __init__(self):
+        # type: () -> None
+        super(S3Uploader, self).__init__()
+        conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
+        bucket_name = settings.S3_AVATAR_BUCKET
+        self.bucket = conn.get_bucket(bucket_name, validate=False)
+
+    def upload_to_s3(self, path, file_data, headers):
+        # type: (Text, binary_type, Optional[Dict[Text, Text]]) -> None
+        key = Key(self.bucket)
+        key.key = path
+        key.set_contents_from_string(force_str(file_data), headers=headers)
+
+    def upload_files(self, response, resized_image, dst_path_id):
+        # type: (Response, binary_type, Text) -> None
+        headers = None  # type: Optional[Dict[Text, Text]]
+        content_type = response.headers.get(str("Content-Type")) or guess_type(dst_path_id)[0]
+        if content_type:
+            headers = {u'Content-Type': content_type}
+        if resized_image:
+            self.upload_to_s3(dst_path_id, resized_image, headers)
+        else:
+            self.upload_to_s3(dst_path_id, response.content, headers)
+        self.upload_to_s3('.'.join((dst_path_id, 'original')), response.content, headers)
+
+def get_uploader():
+    # type: () -> Uploader
+    if settings.LOCAL_UPLOADS_DIR is None:
+        return S3Uploader()
+    return LocalUploader()
+
+
+def upload_emoji_to_storage(apps, schema_editor):
+    # type: (StateApps, DatabaseSchemaEditor) -> None
+    realm_emoji_model = apps.get_model('zerver', 'RealmEmoji')
+    uploader = get_uploader()  # type: Uploader
+    for emoji in realm_emoji_model.objects.all():
+        file_name = uploader.upload_emoji(emoji.realm_id, emoji.img_url, emoji.name)
+        emoji.file_name = file_name
+        emoji.save()
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ('zerver', '0063_realm_description'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='realmemoji',
+            name='file_name',
+            field=models.TextField(db_index=True, null=True),
+        ),
+        migrations.RunPython(upload_emoji_to_storage),
+        migrations.RemoveField(
+            model_name='realmemoji',
+            name='img_url',
+        ),
+    ]
```

Comment:
I've tested it with adding `PNG` and `GIF`(animated and not animated) image URLs to the database and then launched migrations with local and S3 storages(I've used my own S3 storage).